File: src/sessions/parse.py
from __future__ import annotations

import html
import json
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .paths import cursor_cwd_from_path, cursor_parent_id

USER_QUERY_RE = re.compile(r"<user_query>\s*(.*?)\s*</user_query>", re.DOTALL)
TIMESTAMP_RE = re.compile(r"<timestamp>\s*(.*?)\s*</timestamp>", re.DOTALL)
COMMAND_RE = re.compile(r"<command-name>|<command-message>|<command-args>")
CLAUDE_SKIP_TYPES = {
    "mode",
    "file-history-snapshot",
    "attachment",
    "system",
    "progress",
    "queue-operation",
}
CONTENT_SKIP_TYPES = {
    "tool_use",
    "tool_result",
    "thinking",
    "server_tool_use",
    "function_call",
    "function_call_output",
    "custom_tool_call",
    "custom_tool_call_output",
}


@dataclass
class Turn:
    role: str
    text: str


@dataclass
class Parsed:
    id: str
    source: str
    cwd: str | None
    started_at: str | None
    updated_at: str | None
    path: str
    mtime: float
    size: int
    title: str
    parent_id: str | None
    body: str
    turns: list[Turn] = field(default_factory=list)


def parse_file(path: Path, source: str) -> Parsed | None:
    try:
        raw = path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return None
    stat = path.stat()
    parsed = parse_text(raw, source, path)
    if parsed is None:
        return None
    parsed.mtime = stat.st_mtime
    parsed.size = stat.st_size
    parsed.path = str(path)
    if not parsed.updated_at:
        parsed.updated_at = _iso_from_mtime(stat.st_mtime)
    if not parsed.started_at:
        parsed.started_at = parsed.updated_at
    return parsed


def parse_text(raw: str, source: str, path: Path) -> Parsed | None:
    if source == "cursor":
        return _parse_cursor(raw, path)
    if source == "claude":
        return _parse_claude(raw, path)
    if source == "codex":
        return _parse_codex(raw, path)
    raise ValueError(f"unknown source: {source}")


def _parse_cursor(raw: str, path: Path) -> Parsed | None:
    turns: list[Turn] = []
    for obj in _iter_jsonl(raw):
        role = obj.get("role")
        if role not in ("user", "assistant"):
            continue
        message = obj.get("message") or {}
        text = _blocks_text(message.get("content"))
        if role == "user":
            extracted = _cursor_user_text(text)
            if not extracted:
                continue
            text = extracted
        if not text:
            continue
        turns.append(Turn(role, text))
    session_id = path.stem
    title = _first_user_title(turns)
    body = _body_from_turns(turns)
    return Parsed(
        id=session_id,
        source="cursor",
        cwd=cursor_cwd_from_path(path),
        started_at=None,
        updated_at=None,
        path=str(path),
        mtime=0,
        size=0,
        title=title,
        parent_id=cursor_parent_id(path),
        body=body,
        turns=turns,
    )


def _parse_claude(raw: str, path: Path) -> Parsed | None:
    turns: list[Turn] = []
    cwd: str | None = None
    session_id: str | None = None
    parent_id: str | None = None
    started_at: str | None = None
    updated_at: str | None = None
    agent_id: str | None = None
    in_subagents = "subagents" in path.parts

    for obj in _iter_jsonl(raw):
        row_type = obj.get("type")
        if row_type in CLAUDE_SKIP_TYPES:
            continue
        if obj.get("cwd") and not cwd:
            cwd = obj["cwd"]
        if obj.get("sessionId") and not session_id:
            session_id = obj["sessionId"]
        if obj.get("agentId"):
            agent_id = obj["agentId"]
        ts = obj.get("timestamp")
        if isinstance(ts, str):
            if started_at is None:
                started_at = ts
            updated_at = ts
        if obj.get("isMeta"):
            continue
        message = obj.get("message") or {}
        role = row_type if row_type in ("user", "assistant") else message.get("role")
        if role not in ("user", "assistant"):
            continue
        text = _claude_message_text(message.get("content"))
        if role == "user" and _is_junk_user_text(text):
            continue
        if not text:
            continue
        turns.append(Turn(role, text))

    if in_subagents:
        session_id_for_row = agent_id or path.stem
        parent_id = session_id or _parent_dir_id(path)
    else:
        session_id_for_row = session_id or path.stem

    return Parsed(
        id=session_id_for_row,
        source="claude",
        cwd=cwd,
        started_at=started_at,
        updated_at=updated_at,
        path=str(path),
        mtime=0,
        size=0,
        title=_first_user_title(turns),
        parent_id=parent_id,
        body=_body_from_turns(turns),
        turns=turns,
    )


def _parse_codex(raw: str, path: Path) -> Parsed | None:
    turns: list[Turn] = []
    session_id = path.stem
    cwd: str | None = None
    started_at: str | None = None
    updated_at: str | None = None
    parent_id: str | None = None

    for obj in _iter_jsonl(raw):
        ts = obj.get("timestamp")
        if isinstance(ts, str):
            updated_at = ts
        row_type = obj.get("type")
        payload = obj.get("payload") or {}
        if row_type == "session_meta":
            session_id = payload.get("id") or payload.get("session_id") or session_id
            cwd = payload.get("cwd") or cwd
            started_at = payload.get("timestamp") or started_at
            parent_id = (
                payload.get("forked_from_id")
                or payload.get("parent_thread_id")
                or parent_id
            )
            continue
        if row_type != "response_item":
            continue
        if payload.get("type") != "message":
            continue
        role = payload.get("role")
        if role not in ("user", "assistant"):
            continue
        text = _blocks_text(payload.get("content"))
        if role == "user":
            text = _codex_user_text(text)
        if not text:
            continue
        turns.append(Turn(role, text))

    if parent_id == session_id:
        parent_id = None

    return Parsed(
        id=session_id,
        source="codex",
        cwd=cwd,
        started_at=started_at,
        updated_at=updated_at,
        path=str(path),
        mtime=0,
        size=0,
        title=_first_user_title(turns),
        parent_id=parent_id,
        body=_body_from_turns(turns),
        turns=turns,
    )


def format_turns(parsed: Parsed, limit: int = 200_000) -> str:
    lines = [
        f"# id={parsed.id} source={parsed.source} cwd={parsed.cwd or ''}",
        f"# started={parsed.started_at or ''} updated={parsed.updated_at or ''}",
        f"# path={parsed.path}",
    ]
    if parsed.parent_id:
        lines.append(f"# parent_id={parsed.parent_id}")
    lines.append("")
    chunks: list[str] = []
    used = 0
    truncated = False
    for turn in parsed.turns:
        block = f"{turn.role}:\n{turn.text}\n"
        if used + len(block) > limit:
            truncated = True
            remain = limit - used
            if remain > 32:
                chunks.append(block[:remain].rstrip() + "\n")
            break
        chunks.append(block)
        used += len(block)
    text = "\n".join(lines) + "\n" + "\n".join(chunks)
    if truncated:
        text += f"\n[truncated; read {parsed.path}]\n"
    return text.rstrip() + "\n"


def _iter_jsonl(raw: str):
    for line in raw.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(obj, dict):
            yield obj


def _blocks_text(content: Any) -> str:
    if isinstance(content, str):
        return html.unescape(content).strip()
    if not isinstance(content, list):
        return ""
    parts: list[str] = []
    for block in content:
        if isinstance(block, str):
            parts.append(block)
            continue
        if not isinstance(block, dict):
            continue
        btype = block.get("type")
        if btype in CONTENT_SKIP_TYPES:
            continue
        for key in ("text", "input_text", "output_text"):
            value = block.get(key)
            if isinstance(value, str) and value.strip():
                parts.append(value)
                break
    return html.unescape("\n".join(parts)).strip()


def _claude_message_text(content: Any) -> str:
    return _blocks_text(content)


def _cursor_user_text(text: str) -> str:
    match = USER_QUERY_RE.search(text)
    if match:
        return match.group(1).strip()
    cleaned = TIMESTAMP_RE.sub("", text).strip()
    return cleaned


BROWSER_CTX_RE = re.compile(
    r"<in-app-browser-context\b.*?</in-app-browser-context>", re.DOTALL
)
SKILL_DUMP_RE = re.compile(r"<skill\b.*?</skill>", re.DOTALL)


def _codex_user_text(text: str) -> str:
    text = BROWSER_CTX_RE.sub("", text)
    text = SKILL_DUMP_RE.sub("", text)
    marker = "## My request:"
    if marker in text:
        text = text.split(marker, 1)[1]
    return text.strip()


def _is_junk_user_text(text: str) -> bool:
    if not text:
        return True
    if "<local-command-caveat>" in text:
        return True
    if COMMAND_RE.search(text):
        return True
    if text.startswith("<system-reminder>"):
        return True
    return False


def _first_user_title(turns: list[Turn]) -> str:
    for turn in turns:
        if turn.role == "user" and turn.text.strip():
            return _one_line(turn.text, 160)
    return ""


def _body_from_turns(turns: list[Turn]) -> str:
    return "\n\n".join(f"{t.role}: {t.text}" for t in turns)


def _one_line(text: str, limit: int) -> str:
    line = " ".join(text.split())
    if len(line) <= limit:
        return line
    return line[: limit - 1].rstrip() + "…"


def _parent_dir_id(path: Path) -> str | None:
    parts = path.parts
    if "subagents" not in parts:
        return None
    idx = parts.index("subagents")
    if idx == 0:
        return None
    return parts[idx - 1]


def _iso_from_mtime(mtime: float) -> str:
    return datetime.fromtimestamp(mtime, tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

File: internal/transcript/parse.go
package transcript

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"html"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/usuario/sessions/internal/config"
	"github.com/usuario/sessions/internal/model"
)

var userQuery = regexp.MustCompile(`(?s)<user_query>\s*(.*?)\s*</user_query>`)
var timestamp = regexp.MustCompile(`(?s)<timestamp>\s*(.*?)\s*</timestamp>`)
var command = regexp.MustCompile(`<command-name>|<command-message>|<command-args>`)
var browserContext = regexp.MustCompile(`(?s)<in-app-browser-context\b.*?</in-app-browser-context>`)
var skillDump = regexp.MustCompile(`(?s)<skill\b.*?</skill>`)

func ParseFile(path, source string) (*model.Session, error) {
	for attempt := 0; attempt < 2; attempt++ {
		before, err := os.Stat(path)
		if err != nil {
			return nil, err
		}
		file, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		s, parseErr := Parse(file, source, path)
		closeErr := file.Close()
		if parseErr != nil {
			return nil, parseErr
		}
		if closeErr != nil {
			return nil, closeErr
		}
		after, err := os.Stat(path)
		if err != nil {
			return nil, err
		}
		if before.ModTime() == after.ModTime() && before.Size() == after.Size() {
			s.Path = path
			// Keep the operations separate so IEEE-754 rounding matches indexes
			// created by earlier releases.
			s.MTime = float64(after.ModTime().Unix()) + float64(after.ModTime().Nanosecond())/1e9
			s.Size = after.Size()
			fallback := after.ModTime().UTC().Format("2006-01-02T15:04:05Z")
			if s.UpdatedAt == nil {
				s.UpdatedAt = &fallback
			}
			if s.StartedAt == nil {
				s.StartedAt = s.UpdatedAt
			}
			return s, nil
		}
	}
	return nil, fmt.Errorf("file changed while parsing: %s", path)
}

func Parse(r io.Reader, source, path string) (*model.Session, error) {
	s := &model.Session{Source: source, Path: path, ID: strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))}
	reader := bufio.NewReader(r)
	for {
		line, err := reader.ReadBytes('\n')
		if len(bytes.TrimSpace(line)) > 0 {
			var obj map[string]any
			if json.Unmarshal(line, &obj) == nil {
				parseObject(s, obj, path)
			}
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
	}
	if source != "cursor" && source != "claude" && source != "codex" {
		return nil, fmt.Errorf("unknown source: %s", source)
	}
	if source == "cursor" {
		s.CWD = stringPtr(cursorCWD(path))
		s.ParentID = stringPtr(parentDir(path))
	}
	if source == "claude" && strings.Contains(filepath.ToSlash(path), "/subagents/") {
		if s.ParentID == nil {
			s.ParentID = stringPtr(parentDir(path))
		}
	}
	if s.ParentID != nil && *s.ParentID == s.ID {
		s.ParentID = nil
	}
	s.Title = firstTitle(s.Turns)
	var body []string
	for _, t := range s.Turns {
		body = append(body, t.Role+": "+t.Text)
	}
	s.Body = strings.Join(body, "\n\n")
	return s, nil
}

func parseObject(s *model.Session, obj map[string]any, path string) {
	typeName, _ := obj["type"].(string)
	payload, _ := obj["payload"].(map[string]any)
	switch s.Source {
	case "cursor":
		role, _ := obj["role"].(string)
		msg, _ := obj["message"].(map[string]any)
		addTurn(s, role, blockText(msg["content"]), true)
	case "claude":
		if skippedClaude(typeName) {
			return
		}
		if s.CWD == nil {
			s.CWD = mapString(obj, "cwd")
		}
		if sessionID := mapString(obj, "sessionId"); sessionID != nil {
			if strings.Contains(filepath.ToSlash(path), "/subagents/") {
				if s.ParentID == nil {
					s.ParentID = sessionID
				}
			} else if s.ID == strings.TrimSuffix(filepath.Base(path), filepath.Ext(path)) {
				s.ID = *sessionID
			}
		}
		if agent := mapString(obj, "agentId"); agent != nil && strings.Contains(filepath.ToSlash(path), "/subagents/") {
			s.ID = *agent
		}
		if ts := mapString(obj, "timestamp"); ts != nil {
			if s.StartedAt == nil {
				s.StartedAt = ts
			}
			s.UpdatedAt = ts
		}
		if meta, _ := obj["isMeta"].(bool); meta {
			return
		}
		msg, _ := obj["message"].(map[string]any)
		role := typeName
		if role != "user" && role != "assistant" {
			role, _ = msg["role"].(string)
		}
		text := blockText(msg["content"])
		if role == "user" && (strings.Contains(text, "<local-command-caveat>") || command.MatchString(text) || strings.HasPrefix(text, "<system-reminder>")) {
			return
		}
		addTurn(s, role, text, false)
	case "codex":
		if ts, ok := obj["timestamp"].(string); ok {
			s.UpdatedAt = &ts
		}
		if typeName == "session_meta" {
			if v := mapString(payload, "id"); v != nil {
				s.ID = *v
			} else if v := mapString(payload, "session_id"); v != nil {
				s.ID = *v
			}
			if v := mapString(payload, "cwd"); v != nil {
				s.CWD = v
			}
			if v := mapString(payload, "timestamp"); v != nil {
				s.StartedAt = v
			}
			if v := mapString(payload, "forked_from_id"); v != nil {
				s.ParentID = v
			} else if v := mapString(payload, "parent_thread_id"); v != nil {
				s.ParentID = v
			}
			return
		}
		if typeName != "response_item" {
			return
		}
		if kind, _ := payload["type"].(string); kind != "message" {
			return
		}
		role, _ := payload["role"].(string)
		text := blockText(payload["content"])
		if role == "user" {
			text = browserContext.ReplaceAllString(text, "")
			text = skillDump.ReplaceAllString(text, "")
			if _, after, ok := strings.Cut(text, "## My request:"); ok {
				text = after
			}
		}
		addTurn(s, role, text, false)
	}
}
func skippedClaude(t string) bool {
	switch t {
	case "mode", "file-history-snapshot", "attachment", "system", "progress", "queue-operation":
		return true
	}
	return false
}
func addTurn(s *model.Session, role, text string, cursor bool) {
	if role != "user" && role != "assistant" {
		return
	}
	if cursor && role == "user" {
		if m := userQuery.FindStringSubmatch(text); len(m) > 0 {
			text = m[1]
		} else {
			text = timestamp.ReplaceAllString(text, "")
		}
	}
	text = strings.TrimSpace(text)
	if text != "" {
		s.Turns = append(s.Turns, model.Turn{Role: role, Text: text})
	}
}
func blockText(v any) string {
	if text, ok := v.(string); ok {
		return strings.TrimSpace(html.UnescapeString(text))
	}
	blocks, ok := v.([]any)
	if !ok {
		return ""
	}
	var parts []string
	for _, raw := range blocks {
		if text, ok := raw.(string); ok {
			parts = append(parts, text)
			continue
		}
		b, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		kind, _ := b["type"].(string)
		switch kind {
		case "tool_use", "tool_result", "thinking", "server_tool_use", "function_call", "function_call_output", "custom_tool_call", "custom_tool_call_output":
			continue
		}
		for _, key := range []string{"text", "input_text", "output_text"} {
			if text, ok := b[key].(string); ok && strings.TrimSpace(text) != "" {
				parts = append(parts, text)
				break
			}
		}
	}
	return strings.TrimSpace(html.UnescapeString(strings.Join(parts, "\n")))
}
func mapString(m map[string]any, key string) *string {
	if v, ok := m[key].(string); ok && v != "" {
		return &v
	}
	return nil
}
func stringPtr(v string) *string {
	if v == "" {
		return nil
	}
	return &v
}
func firstTitle(turns []model.Turn) string {
	for _, t := range turns {
		if t.Role == "user" && strings.TrimSpace(t.Text) != "" {
			line := strings.Join(strings.Fields(t.Text), " ")
			if utf8.RuneCountInString(line) > 160 {
				r := []rune(line)
				line = strings.TrimSpace(string(r[:159])) + "…"
			}
			return line
		}
	}
	return ""
}
func parentDir(path string) string {
	parts := strings.Split(filepath.ToSlash(path), "/")
	for i, p := range parts {
		if p == "subagents" && i > 0 {
			return parts[i-1]
		}
	}
	return ""
}
func cursorCWD(path string) string {
	rootAbs, _ := filepath.Abs(config.CursorRoot())
	pathAbs, _ := filepath.Abs(path)
	rel, err := filepath.Rel(rootAbs, pathAbs)
	if err != nil || strings.HasPrefix(rel, "..") {
		return ""
	}
	slug := strings.Split(filepath.ToSlash(rel), "/")[0]
	tokens := strings.Split(slug, "-")
	current := string(filepath.Separator)
	for i := 0; i < len(tokens); {
		found := -1
		for j := len(tokens); j > i; j-- {
			candidate := filepath.Join(current, strings.Join(tokens[i:j], "-"))
			if st, e := os.Stat(candidate); e == nil && st.IsDir() {
				found = j
				current = candidate
				break
			}
		}
		if found < 0 {
			return filepath.Join(current, strings.Join(tokens[i:], "-"))
		}
		i = found
	}
	return current
}

func Format(s *model.Session, limit int) string {
	lines := []string{fmt.Sprintf("# id=%s source=%s cwd=%s", s.ID, s.Source, value(s.CWD)), fmt.Sprintf("# started=%s updated=%s", value(s.StartedAt), value(s.UpdatedAt)), "# path=" + s.Path}
	if s.ParentID != nil {
		lines = append(lines, "# parent_id="+*s.ParentID)
	}
	out := strings.Join(lines, "\n") + "\n\n"
	used := 0
	truncated := false
	var blocks []string
	for _, t := range s.Turns {
		block := t.Role + ":\n" + t.Text + "\n"
		n := utf8.RuneCountInString(block)
		if used+n > limit {
			truncated = true
			remain := limit - used
			if remain > 32 {
				r := []rune(block)
				blocks = append(blocks, strings.TrimRight(string(r[:remain]), " \n\t")+"\n")
			}
			break
		}
		blocks = append(blocks, block)
		used += n
	}
	out += strings.Join(blocks, "\n")
	if truncated {
		out += "\n[truncated; read " + s.Path + "]\n"
	}
	return strings.TrimRight(out, "\n") + "\n"
}
func value(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

var _ = time.RFC3339
